### DISCUSSION TITLE
List and view performance

### DISCUSSION THREAD

**Discussion by lorenzocattaneo** [pocketbase/pocketbase#2820] [Q&A]
Hello, 

I'm trying do some benchmarks to decide if Pocketbase would be appropriate for my project. I'm having some problems with listing records and rules, that performance seems to be degrading quite fast with growing collections. Of course I could very well be doing things wrong, if anyone can help me out it would be appreciated.

Here's what I've got:
-I've tested on a Hetzner CPX21 machine
-I've tested on my local machine (Apple M1 Max, 32GB RAM)
-tests are run with wrk. Benchmark results are, in proportion, the same between the 2 environments. I'll be providing my local machine results as an example.
-for each system I've downloaded the appropriate pre-built version from the github releases
-I've created organisations collection with a  plain text field (name) and a multiple relation towards users.
-I've created a messages collection with a plain text message field and 2 single relations, one for the organisation and one for the user
-I've also tried to run pocketbase with debug on, and in each scenario 99%of queries indicate 0.0ms, with some occasional query being 10ms (the slowest)

Initially I've removed all API Rules and created one user, organisation and a single message manually.

```
wrk  --latency  -c 10 -d 10s -t 2 http://localhost:8090/api/collections/messages/records\?page\=1\&perPage\=5
Running 10s test @ http://localhost:8090/api/collections/messages/records?page=1&perPage=5
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.17ms    1.61ms  49.05ms   93.97%
    Req/Sec     5.36k     1.55k    7.02k    86.63%
  Latency Distribution
     50%  791.00us
     75%    1.19ms
     90%    2.08ms
     99%    6.69ms
  107665 requests in 10.10s, 67.97MB read
Requests/sec:  10655.31
Transfer/sec:      6.73MB
```

Then I enabled this List rule: `org.users.id = @request.auth.id`

```
wrk -H "Authorization: XXX"--latency  -c 100 -d 10s -t 2 http://localhost:8090/api/collections/messages/records\?page\=1\&perPage\=5
Running 10s test @ http://localhost:8090/api/collections/messages/records?page=1&perPage=5
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     7.14ms    2.19ms  27.18ms   76.21%
    Req/Sec   705.94     26.34   767.00     71.00%
  Latency Distribution
     50%    6.88ms
     75%    8.19ms
     90%    9.65ms
     99%   14.52ms
  14069 requests in 10.02s, 8.88MB read
Requests/sec:   1404.63
Transfer/sec:      0.89MB
```

Performance is much slower already but still ok.

Then I've added a second organisation and user, and added about 10k messages to the collection. each user has 5k messages.

Just for reference, I got ~ 1600req/s with inserts

Here's the same tests, without the list rule first, and with it enabled after:
```
wrk -H "Authorization: XXX" --latency  -c 10 -d 10s -t 2 http://localhost:8090/api/collections/messages/records\?page\=1\&perPage\=5
Running 10s test @ http://localhost:8090/api/collections/messages/records?page=1&perPage=5
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     4.67ms    6.44ms  85.41ms   91.28%
    Req/Sec     1.58k     1.13k    4.54k    72.00%
  Latency Distribution
     50%    2.68ms
     75%    5.02ms
     90%    9.99ms
     99%   34.48ms
  31493 requests in 10.02s, 62.17MB read
Requests/sec:   3143.01
Transfer/sec:      6.20MB
```

```
wrk -H "Authorization: XXX" --latency  -c 10 -d 10s -t 2 http://localhost:8090/api/collections/messages/records\?page\=1\&perPage\=5
Running 10s test @ http://localhost:8090/api/collections/messages/records?page=1&perPage=5
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   353.67ms   19.83ms 440.80ms   83.27%
    Req/Sec    14.83      7.29    30.00     82.31%
  Latency Distribution
     50%  354.53ms
     75%  362.46ms
     90%  371.81ms
     99%  397.76ms
  281 requests in 10.08s, 570.23KB read
Requests/sec:     27.89
Transfer/sec:     56.59KB
```

I've run each test multiple times and the results are always consistent. The drop in performance is very noticeable in both situations.

I've done a final test with 50k items in the collection:

```
wrk -H "Authorization: XXX" --latency  -c 10 -d 10s -t 2 http://localhost:8090/api/collections/messages/records\?page\=1\&perPage\=5
Running 10s test @ http://localhost:8090/api/collections/messages/records?page=1&perPage=5
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     4.74ms    3.19ms  57.35ms   93.20%
    Req/Sec     1.14k   166.56     1.42k    78.00%
  Latency Distribution
     50%    3.92ms
     75%    4.99ms
     90%    6.79ms
     99%   18.62ms
  22618 requests in 10.01s, 44.67MB read
Requests/sec:   2258.70
Transfer/sec:      4.46MB
```

```
wrk -H "Authorization: XXX" --latency  -c 10 -d 10s -t 2 http://localhost:8090/api/collections/messages/records\?page\=1\&perPage\=5
Running 10s test @ http://localhost:8090/api/collections/messages/records?page=1&perPage=5
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.92s    35.15ms   1.99s    69.77%
    Req/Sec    10.91     12.62    39.00     81.82%
  Latency Distribution
     50%    1.92s
     75%    1.94s
     90%    1.97s
     99%    1.99s
  50 requests in 10.08s, 101.51KB read
  Socket errors: connect 0, read 0, write 0, timeout 7
Requests/sec:      4.96
Transfer/sec:     10.07KB
```

Just out of curiosity I've also tried to add a custom route where I'm manually running the query with this code:

```Go
e.Router.AddRoute(echo.Route{
			Method: http.MethodGet,
			Path:   "/demo",
			Handler: func(c echo.Context) error {
				var records []*Message
				err := app.Dao().DB().
					Select("message").
					From("messages").Where(dbx.NewExp("organisation={:organisation}", dbx.Params{
					"organisation": c.Request().Header.Get("X-Demo-Header"),
				})).
					OrderBy("created").
					Limit(50).
					All(&records)

				return c.JSON(200, records)
			},
			Middlewares: []echo.MiddlewareFunc{
				apis.ActivityLogger(app),
				apis.RequireRecordAuth(),
			},
		})
```

which produced these results on the 50k items collection:

```
wrk -H "Authorization: XXX" --latency  -c 10 -d 10s -t 2 http://localhost:8090/demo
Running 10s test @ http://localhost:8090/demo
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   815.57us    1.28ms  27.01ms   94.10%
    Req/Sec     8.38k     1.79k   10.78k    64.00%
  Latency Distribution
     50%  502.00us
     75%  828.00us
     90%    1.47ms
     99%    5.99ms
  166920 requests in 10.00s, 37.25MB read
Requests/sec:  16689.82
Transfer/sec:      3.72MB
```

Of course this version is much less configurable and does only one specific thing, I was merely trying to understand how much each part of the framework impacts performance.

Does anyone know if this can be improved? Is there a smarter way to do this that retains good performance?

Thanks!

**Comment by ganigeorgiev** [comment]
Thank you for sharing your benchmark results. Official testbed and benchmarks will be available soon (_I'm still not sure what would be the correct way to do it to avoid creating some useless artificial benchmark_).

Keep in mind that the executed queries with and without rules are different and the collection API rules will always have some cost (especially when doing nested relations lookups).

To understand why the performance degrades with the API rule enabled, let's try to analyze `org.users.id = @request.auth.id`.

The above initially will produce 2 `JOIN`s with the `org` and the `users` collections (_actually it will be 3 JOINS with the `json_each` for the users array value, but with small number of selected items that extra join is negligible in most cases_).
Additionally, because in your case `org.users` is a multiple relation, the `=` operator will be match-all, aka. it generates a **subuquery** where it will require **every** `org.users` to be equal to the current authenticated user id, which will most likely always fail because a single user can be selected only once and what you want is probably the `?=` operator, aka. "any/at least one of" the selected `org.users` to have an id equal the authenticated user.

To optmize the above query, you can try to:

- replace `=` with `?=`
- add individual indexes for the relation fields (_I'll consider in future version to suggest doing that with the relation field creation_)

If the above doesn't help, could you create a repo and share it so that I can try to run it locally?

**Comment by lorenzocattaneo** [comment]
The relations were already indexed, I forgot to write that down!

I've tried to replace `=` with `?=` as you suggested, there's some performance improvement but not much. With 50k messages, it went from ~5req/s to ~10req/s. 
This seems reasonable since there is still only a single user per organisation in the test, so even a less efficient query doesn't have much more to do.

I've tried a different approach by making a `organisation_user` collection, where each row has a single relation to the `organisation` and `user`, and setting `organisation.id = @collection.organisation_user.organisation.id && @request.auth.id = @collection.organisation_user.user.id` as a rule. That made more of a difference, and got me ~110req/s

Anyway, I've prepared a repo as you suggested: [https://github.com/lorenzocattaneo/pocketbase-benchmark-demo](https://github.com/lorenzocattaneo/pocketbase-benchmark-demo)

Admin user is admin@demo.com 
All users (admin included) have password admin12345

In the backups you can find collections with 1, 5k and 25k messages per user, so you just need to adjust the api rules to test it out.

**Reply by dschissler** [reply]
@lorenzocattaneo Have you tried creating custom views?

@ganigeorgiev Will the API Filter rules use the existing joins or does it need to join all over again?

**Reply by ganigeorgiev** [reply]
> Will the API Filter rules use the existing joins or does it need to join all over again?

I don't understand what do you mean by "existing joins". Every time when you have a relation field lookup (eg. `someRelField.prop`) it will add a `JOIN` statement to the base query.

**Reply by dschissler** [reply]
I mean that if you have a SQL custom view that is already doing these joins, say like joining five tables, then if the API Filter is accessing those same tables does it need to do a join all over again (even though its in the SQL already)?  I mean is there any possible efficiency that could be achieved there that would be faster than having SQLite optimize it?

**Reply by ganigeorgiev** [reply]
No, the API rules don't modify the existing view query. The view query is treated as a table, aka. practically the base query for a view collection is `SELECT ... FROM (view) ...` and the API rule joins are appended to that base query.

>  I mean is there any possible efficiency that could be achieved there that would be faster than having SQLite optimize it?

I'm not sure, I haven't tested it, but this internally will be a lot more code involving as it will require parsing the SQL statement.

The next couple days I'll try to prepare a testbed to have some basis for comparison and will see if we can fine-tune some of the generated queries.

**Reply by lorenzocattaneo** [reply]
@dschissler as far as I know in sqlite views are not stored so the query is executed every time and as pointed out, the query Pocketbase generates already appends the appropriate joins to the base query, so the result should be similar. At least I think so?
Regardless, sqlite views cannot contain parameters so at least some condition would have to be added to the view query to filter by the current organisation.

@ganigeorgiev I've tried to gather some more scenarios by simplifying the query as much as possible to make sure it runs fast and see what changes.

I've added an `active_organisation` relation to the user directly, and changed the rule to `organisation = @request.auth.active_organisation`

It products this query
```
SELECT DISTINCT `messages`.* FROM `messages` LEFT JOIN `users` `__auth_users` ON `__auth_users`.`id`='pf9ow9hh4gsm6ip' WHERE COALESCE([[messages.organisation]], '') = COALESCE([[__auth_users.active_organisation]], '') LIMIT 30

```
which is both fast and simple and gets me to about 180req/s. Since the JWT does not include the `active_organisation` field a join is necessary.

So I've changed to rule to `user = @request.auth.id` which gives me this:

```
SELECT `messages`.* FROM `messages` WHERE [[messages.user]] = 'pf9ow9hh4gsm6ip' LIMIT 5
```
and around 580req/s

The exact same rule on a 1 message per user collection nets ~2200req/s

I'm not sure if this query is just executed or run as a prepared statement or if some indexing could be done better on my end. Hope this helps.

If I have some time in the next few days I might try to investigate the codebase and see if I can help out. No promises

**Reply by ganigeorgiev** [reply]
_@lorenzocattaneo Just to let you know that I've added a minor optimization to the `COUNT` search query in [v0.16.7 release](https://github.com/pocketbase/pocketbase/releases/tag/v0.16.7) that may improve further the performance in your case with the `getList()` calls._
